cli/sql: report the cluster ID and server version upon new connections

knz opened the pull request:
### sql: expose the node ID of the gateway node as a read-only session variable

The new variable is called `node_id`.

This makes it more convenient than `SELECT node_id FROM
crdb_internal.node_build_info LIMIT 1`.

Requested/suggested by @jseldess.

Fixes #14310.

### cli/sql: report the cluster ID and server version upon new connections

This patch ensures the client prints out the cluster ID and
server version, and if different from the server version, also the
client version, both upon the initial connection and upon reconnects
*if they change*.

For example:

    $ ./cockroach sql
    # Welcome to the cockroach SQL interface.
    # All statements must be terminated by a semicolon.
    # To exit: CTRL + D.
    #
    # Server version: CockroachDB CCL 0da9d8f-dirty (linux amd64, built 2017/06/23 19:50:15, go1.8) (same version as client)
    # Cluster ID: fe26b6f3-3153-4c75-a7fb-3027c327e5ab
    #
    # Enter \? for a brief introduction.
    #
    root@localhost:26257/>

Then upon a reconnect, if the node ID changes:

    root@localhost:26257/> SELECT 1;
    driver: bad connection
    connection lost; opening new connection: all session settings will be lost
    root@localhost:26257/>

If the server version has changed:

    root@localhost:26257/> select 1;
    driver: bad connection
    connection lost; opening new connection: all session settings will be lost
    # Client version: CockroachDB CCL a1b6734-dirty (linux amd64, built 2017/06/23 19:14:15, go1.8)
    # Server version: CockroachDB CCL a1b6734-dirty (linux amd64, built 2017/06/23 19:15:00, go1.8)
    root@localhost:26257/>

If the cluster ID has changed:

        root@localhost:26257/> SELECT 1;
        driver: bad connection
        connection lost; opening new connection: all session settings will be lost
        error retrieving the transaction status: the cluster ID has changed!
        Previous ID: b8d0cc65-3a71-4db8-bcda-6fa86206ebb0
        New ID: fe26b6f3-3153-4c75-a7fb-3027c327e5ab
        root@localhost:26257/ ?>

Supersedes #16615.


cc @mberhault

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16703)
<!-- Reviewable:end -->

knz commented:
Marc I'm assinging to you because you had an interest in the future and Matt is out this week. Feel free to de-assign if this feels out of your league.

mberhault commented on pkg/cli/sql_util.go:
```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
+		// actually changes. This is sign of a potentially very bad load
+		// balancer configuration.
+		if c.clusterID != "" {
+			fmt.Print("\n# WARNING: the cluster ID has changed!\n\n")
```
This isn't enough, we should error out if reconnecting to a different cluster. I'd even be happy with a hard `log.Fatal`

mberhault commented on pkg/cli/sql_util.go:
```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
+	// and it has changed since the last connection.
+	if newNodeID != c.nodeID && newNodeID != "" {
+		c.nodeID = newNodeID
+		fmt.Println("# Node ID:", newNodeID)
```
I still don't think this is necessary. We should document the contents of `crdb_internal` for people who care, but we definitely don't need to log this for most users.

knz replied:
ok, removed. Added a session variable as recommended/suggested by Jesse.

mberhault commented on pkg/cli/sql_util.go:
```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
+
+		serverVersionShort := newServerBuildInfo.Short()
+		isSame := ""
+		if serverVersionShort == clientVersionShort {
```
I would limit the "sameness" check to the tag. We don't care if they're different platforms/build tools/timestamps. The important bit is the wire protocol.

knz commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql_util.go, line 159 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-159:-KnZTmveaZKch9QZhDZR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L159)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I would limit the "sameness" check to the tag. We don't care if they're different platforms/build tools/timestamps. The important bit is the wire protocol.
</blockquote></details>

Done.

---

*[pkg/cli/sql_util.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-176:-KnZTnV93Ft7wYnsjuXv:b-8k12a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L176)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

This isn't enough, we should error out if reconnecting to a different cluster. I'd even be happy with a hard `log.Fatal`
</blockquote></details>

`log.Fatal` is not suitable: it's common (at least in my workflow, but I know I'm not the only one) to leave an interactive SQL shell open across test cluster resets. 

Instead I changed to use an error, which prevents the query from being executed.

---

*[pkg/cli/sql_util.go, line 187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-187:-KnZVNEP1hnTcFQSlLoq:b-u5o6sl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L187)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I still don't think this is necessary. We should document the contents of `crdb_internal` for people who care, but we definitely don't need to log this for most users.
</blockquote></details>

That's fine by me. @jseldess can you confirm that #14310 would be suitably addressed by recommending a SELECT query into `crdb_internal.node_build_info` instead of having the CLI shell report the node ID in interactive sessions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[pkg/cli/sql_util.go, line 187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703#-KnLcDt8c4yGhIt3j-Tt-r3-187:-KnZpmJV5Z5nuhLihdYT:bkg0zmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2b9bc6e8b4053edf2ca0d2af1f238ae5023a95/pkg/cli/sql_util.go#L187)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That's fine by me. @jseldess can you confirm that #14310 would be suitably addressed by recommending a SELECT query into `crdb_internal.node_build_info` instead of having the CLI shell report the node ID in interactive sessions?
</blockquote></details>

Sure, I'm also fine leaving it out. We already do recommend querying that internal table in a few places, for example: https://www.cockroachlabs.com/docs/manual-deployment.html#step-6-test-load-balancing. In the referenced issue, we also discussed a session variable for getting the node ID...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16703)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: sort the session variables in the source code

knz committed:
sql: expose the node ID of the gateway node as a read-only session variable

The new variable is called `node_id`.

This makes it more convenient than `SELECT node_id FROM
crdb_internal.node_build_info LIMIT 1`.

Requested/suggested by @jseldess.

knz committed:
cli/sql: move getServerValue to sql_util

knz committed:
cli/sql: enhance the introductory comment

This patch enhances the message printed upon startup by `cockroach
sql` in interactive sessions.

Before this patch:

    $ ./cockroach sql
    # Welcome to the cockroach SQL interface.
    # All statements must be terminated by a semicolon.
    # To exit: CTRL + D.
    #
    root@localhost:26257/>

After this patch:

    $ ./cockroach sql
    # Welcome to the cockroach SQL interface.
    # All statements must be terminated by a semicolon.
    # To exit: CTRL + D.
    #
    # Enter \? for a brief introduction.
    #
    root@localhost:26257/>

Or in case of connection errors:

Before this patch:

    $ ./cockroach sql --host=nonexistent
    Error: unable to connect or connection lost.
    [...]
    Failed running "sql"

After this patch:

    # Welcome to the cockroach SQL interface.
    # All statements must be terminated by a semicolon.
    # To exit: CTRL + D.
    #
    Error: unable to connect or connection lost.
    [...]
    Failed running "sql"

(Why: this way, any reporting done about the SQL session while the
connection is being set up will appear after the welcome message.)

knz committed:
cli/sql: report the cluster ID, node ID and server version upon new connections

This patch ensures the client prints out the cluster ID and server
version, and if different from the server version, also the client
version, both upon the initial connection and upon reconnects *if they
change*.

For example:

    $ ./cockroach sql
    # Welcome to the cockroach SQL interface.
    # All statements must be terminated by a semicolon.
    # To exit: CTRL + D.
    #
    # Server version: CockroachDB CCL 0da9d8f-dirty (linux amd64, built 2017/06/23 19:50:15, go1.8) (same version as client)
    # Cluster ID: fe26b6f3-3153-4c75-a7fb-3027c327e5ab
    #
    # Enter \? for a brief introduction.
    #
    root@localhost:26257/>

If the server version has changed:

    root@localhost:26257/> select 1;
    driver: bad connection
    connection lost; opening new connection: all session settings will be lost
    # Client version: CockroachDB CCL a1b6734-dirty (linux amd64, built 2017/06/23 19:14:15, go1.8)
    # Server version: CockroachDB CCL a1b6734-dirty (linux amd64, built 2017/06/23 19:15:00, go1.8)
    root@localhost:26257/>

If the cluster ID has changed:

    root@localhost:26257/> SELECT 1;
    driver: bad connection
    connection lost; opening new connection: all session settings will be lost
    error retrieving the transaction status: the cluster ID has changed!
    Previous ID: b8d0cc65-3a71-4db8-bcda-6fa86206ebb0
    New ID: fe26b6f3-3153-4c75-a7fb-3027c327e5ab
    root@localhost:26257/ ?>

mberhault commented on pkg/cli/sql_util.go:
```diff
@@ -69,12 +79,160 @@ func (c *sqlConn) ensureConn() error {
+		// Read the node_build_info table as an array of strings.
+		rowVals, err := getAllRowStrings(rows, true /* showMoreChars */)
+		if err != nil || len(rowVals) == 0 || len(rowVals[0]) != 3 {
+			fmt.Fprintln(stderr, "error while rerieving the server's version")
```
s/rerieving/retrieving/

knz replied:
Done.

mberhault reviewed (approved):
LGTM, though you haven't pushed your changes to the last round of comments.

knz commented:
I did push the changes! They are updated at the source branch https://github.com/knz/cockroach/tree/20170623-cli-ver-reporting

somehow GH is not picking it up here?!

knz commented:
ok it's picked up now :)

knz commented:
ok that went through! TFYRs
